rootless: btrfs: annotate error with human-readable hint string

AkihiroSuda opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Add hints for `Failed to destroy btrfs snapshot <DIR> for <ID>: operation not permitted` error on rootless

Related to issue #41762

**- How I did it**

Wrap the error when running in a user namespace.

**- How to verify it**

```console
$ dockerd-rootless.sh --data-root /mnt/btrfs/varlibdocker
```

```console
$ docker --context=rootless run --rm hello-world
Hello from Docker!
...
ERRO[0006] Error waiting for container: container 59753471c3be3525b46678702e52445aa596da01ccd19e502fde9ff14ec1cfe9: driver "btrfs" failed to remove root filesystem: failed to delete subvolume without root (hint: remount btrfs on "user_subvol_rm_allowed" option, or change the storage driver to "fuse-overlayfs"): Failed to destroy btrfs snapshot /mnt/btrfs/varlibdocker/btrfs/subvolumes for 8a09ee12fd5cb205dfd554a189a389645bb38184703874cd92745690868d845a: operation not permitted 
```

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
rootless: btrfs: annotate error with human-readable hint string

**- A picture of a cute animal (not mandatory but encouraged)**

:penguin:

AkihiroSuda commented on daemon/graphdriver/btrfs/btrfs.go:
```diff
@@ -633,6 +634,10 @@ func (d *Driver) Remove(id string) error {
+		if sys.RunningInUserNS() && strings.Contains(err.Error(), "operation not permitted") {
+			// https://github.com/moby/moby/issues/41762
+			err = errors.Wrap(err, "failed to delete subvolume without root (hint: remount btrfs on \"user_subvol_rm_allowed\" option, or change the storage driver to \"fuse-overlayfs\")")
+		}
```
Note: proper fix is being implemented in https://github.com/moby/moby/pull/42203 , but requires kernel >= 4.18.
So I think we can have both PRs.

thaJeztah commented on daemon/graphdriver/btrfs/btrfs.go:
```diff
@@ -633,6 +634,10 @@ func (d *Driver) Remove(id string) error {
 	d.updateQuotaStatus()
 
 	if err := subvolDelete(d.subvolumesDir(), id, d.quotaEnabled); err != nil {
+		if sys.RunningInUserNS() && strings.Contains(err.Error(), "operation not permitted") {
```
Would we be able to get an error-type instead of string matching?

AkihiroSuda replied:
Doesn't seem straightforward https://github.com/moby/moby/blob/8e29dd734d20ddf503fb974190fac42cca71aa42/daemon/graphdriver/btrfs/btrfs.go#L267

thaJeztah replied:
Isn't it coming from https://github.com/keybase/client/blob/master/go/vendor/golang.org/x/sys/unix/zerrors_linux_amd64.go#L2921 ?

Was thinking something like `errors.Is(err, os.ErrPermission)`

AkihiroSuda replied:
That requires replacing `%v` with `%w`

tonistiigi reviewed (approved):
I don't think this should be cherry-picked

AkihiroSuda committed:
btrfs: annotate error with human-readable hint string

Add hints for "Failed to destroy btrfs snapshot <DIR> for <ID>: operation not permitted" on rootless

Related to issue 41762

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda commented:
rebased

thaJeztah commented on daemon/graphdriver/btrfs/btrfs.go:
```diff
@@ -634,6 +635,9 @@ func (d *Driver) Remove(id string) error {
 	if err := subvolDelete(d.subvolumesDir(), id, d.quotaEnabled); err != nil {
 		if d.quotaEnabled {
+			if sys.RunningInUserNS() && strings.Contains(err.Error(), "operation not permitted") {
+				err = errors.Wrap(err, "failed to delete subvolume without root (hint: remount btrfs on \"user_subvol_rm_allowed\" option, or update the kernel to >= 4.18, or change the storage driver to \"fuse-overlayfs\")")
```
Perhaps use backticks for this one;

```suggestion
				err = errors.Wrap(err, `failed to delete subvolume without root (hint: remount btrfs on "user_subvol_rm_allowed" option, or update the kernel to >= 4.18, or change the storage driver to "fuse-overlayfs")`)
```

AkihiroSuda replied:
done

thaJeztah commented on daemon/graphdriver/btrfs/btrfs.go:
```diff
@@ -26,6 +26,7 @@ import (
 	"sync"
 	"unsafe"
 
+	"github.com/containerd/containerd/sys"
```
Could you do another rebase, and update this to use `"github.com/containerd/containerd/pkg/userns"` ?

```suggestion
	"github.com/containerd/containerd/pkg/userns"
```


Full diff (with my other suggestion);

```patch
diff --git a/daemon/graphdriver/btrfs/btrfs.go b/daemon/graphdriver/btrfs/btrfs.go
index 677dc7aed4..0d0ee3a207 100644
--- a/daemon/graphdriver/btrfs/btrfs.go
+++ b/daemon/graphdriver/btrfs/btrfs.go
@@ -26,7 +26,7 @@ import (
        "sync"
        "unsafe"

-       "github.com/containerd/containerd/sys"
+       "github.com/containerd/containerd/pkg/userns"
        "github.com/docker/docker/daemon/graphdriver"
        "github.com/docker/docker/pkg/containerfs"
        "github.com/docker/docker/pkg/idtools"
@@ -601,8 +601,8 @@ func (d *Driver) Remove(id string) error {

        if err := subvolDelete(d.subvolumesDir(), id, d.quotaEnabled); err != nil {
                if d.quotaEnabled {
-                       if sys.RunningInUserNS() && strings.Contains(err.Error(), "operation not permitted") {
-                               err = errors.Wrap(err, "failed to delete subvolume without root (hint: remount btrfs on \"user_subvol_rm_allowed\" option, or update the kernel to >= 4.18, or change the storage driver to \"fuse-overlayfs\")")
+                       if userns.RunningInUserNS() && strings.Contains(err.Error(), "operation not permitted") {
+                               err = errors.Wrap(err, `failed to delete subvolume without root (hint: remount btrfs on "user_subvol_rm_allowed" option, or update the kernel to >= 4.18, or change the storage driver to "fuse-overlayfs")`)
                        }
                        return err
                }
```

AkihiroSuda replied:
done

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
Failure is unrelated; merging
